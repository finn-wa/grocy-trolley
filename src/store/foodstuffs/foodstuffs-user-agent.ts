import { getEnv } from "env";
import { access } from "fs/promises";
import { Headers, Response as NodeResponse } from "node-fetch";
import { Browser, BrowserContext, JSHandle, Page } from "playwright";
import { APPLICATION_JSON, headers } from "utils/headers-builder";
import { Logger, prettyPrint } from "utils/logger";
import { PAKNSAVE_URL } from "./foodstuffs.model";

export class FoodstuffsUserAgent {
  protected readonly logger = new Logger(this.constructor.name);
  private readonly loginDetails: LoginDetails;
  private context?: BrowserContext;
  private page?: Page;

  // TODO: support creating a user agent without logging in. for unauthenticated search requests
  constructor(
    private readonly browser: Browser,
    loginDetails?: LoginDetails,
    private readonly storageStatePath = "src/resources/cache/playwright.json"
  ) {
    if (loginDetails) {
      this.loginDetails = loginDetails;
    } else {
      const { PAKNSAVE_EMAIL: email, PAKNSAVE_PASSWORD: password } = getEnv();
      this.loginDetails = { email, password };
    }
  }

  /**
   * Performs a fetch request as a logged in user on the PAK'n'SAVE website.
   * @param method Request method
   * @param url Request URL
   * @param headers Request headers
   * @param body Request body
   * @returns Request response
   */
  async fetchAsUser(
    method: string,
    url: string,
    headers?: Headers,
    body?: any
  ): Promise<NodeResponse> {
    const page = await this.getPage();
    return this.fetchWithPage(page, method, url, headers, body);
  }

  /**
   * Performs a fetch request from the browser using Playwright. Necessary
   * because Cloudflare now blocks requests that are not sent from a browser.
   * WARNING: not every method on the returned Response is callable, see
   * {@link FoodstuffsRequest}.
   * @param page Playwright page instance to use to perform the request
   * @param method Request method
   * @param url Request URL
   * @param headers Request headers
   * @param body Request body
   * @returns Request response
   */
  private async fetchWithPage(
    page: Page,
    method: string,
    url: string,
    headers?: Headers,
    body?: BodyInit | any
  ): Promise<NodeResponse> {
    this.logger.debug(`${method} ${url}`);
    if (headers) {
      this.logger.trace(headers);
    }
    if (body) {
      this.logger.trace(body);
    }
    const contentType = headers?.get("content-type");
    if (contentType === APPLICATION_JSON && body) {
      body = JSON.stringify(body);
    }
    const responseHandle = await page.evaluateHandle(
      async ({ url, method, headers, body }) =>
        fetch(url, {
          credentials: "include",
          referrer: "https://www.paknsave.co.nz/shop",
          mode: "cors",
          method,
          headers,
          body,
        }),
      { url, method, headers, body }
    );
    // Response is not serialisable, so we must serialise it
    const response = await responseHandle.evaluate((response, props) => {
      const responseJson: Record<string, any> = Object.fromEntries(
        props.map((key) => [key, response[key]])
      );
      responseJson.headers = Object.fromEntries(responseJson.headers.entries());
      return responseJson as SerialisedResponse;
    }, SERIALISABLE_RESPONSE_PROPS);
    if (!response.ok) {
      throw new Error(`Response not OK: ${prettyPrint(response)}`);
    }
    this.logger.trace("Response: " + prettyPrint(response));
    return new FoodstuffsResponse(responseHandle, response);
  }

  /**
   * Returns the login page. Creates a new page and logs in if needed.
   * @returns Login page
   */
  private async getPage(): Promise<Page> {
    if (this.page) {
      return this.page;
    }
    const context = await this.getContext();
    const page = await context.newPage();
    await page.goto(PAKNSAVE_URL + "shop");
    // Check if already logged in
    try {
      await this.fetchWithPage(
        page,
        "GET",
        `${PAKNSAVE_URL}/CommonApi/Account/GetUserProfile`,
        headers().acceptJson().build()
      );
    } catch (error) {
      await page.click('button[id="login-form"]');
      await page.fill('input[id="login-email"]', this.loginDetails.email);
      await page.fill('input[id="login-password"]', this.loginDetails.password);
      await page.click("button.login-form-submit");
      // page refreshes after submission
      await page.waitForLoadState("networkidle");
      // search box seems to pop in last
      await page.locator('input[aria-label="Search products"]').waitFor();
      // save logged in state
      await page.context().storageState({ path: this.storageStatePath });
    }
    this.page = page;
    return this.page;
  }

  /**
   * Returns the browser context with cached storage state (if found).
   * Creates a new browser context if needed.
   * @returns Browser context
   */
  private async getContext(): Promise<BrowserContext> {
    if (!this.context) {
      try {
        await access(this.storageStatePath);
        // No error means that cached storageState exists
        this.context = await this.browser.newContext({ storageState: this.storageStatePath });
      } catch (error) {
        this.logger.info("No storageState found, creating new context");
        this.context = await this.browser.newContext();
      }
    }
    return this.context;
  }
}

export interface LoginDetails {
  email: string;
  password: string;
}

export interface FoodstuffsUserProfile {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  agreedToMarketing: boolean;
  agreedToTermsAndConditions: boolean;
  clubCardUser: boolean;
}

export interface LoginResponse {
  success: boolean;
  userProfile: FoodstuffsUserProfile;
}

/**
 * Response backed by a Playwright JSHandle browser fetch Response. json()
 * and text() extraction methods are available.
 */
class FoodstuffsResponse implements NodeResponse {
  readonly bodyUsed: boolean;
  readonly headers: Headers;
  readonly ok: boolean;
  readonly redirected: boolean;
  readonly status: number;
  readonly statusText: string;
  readonly type: ResponseType;
  readonly url: string;

  constructor(
    private readonly responseHandle: JSHandle<Response>,
    responseJson: SerialisedResponse
  ) {
    this.bodyUsed = responseJson.bodyUsed;
    this.ok = responseJson.ok;
    this.redirected = responseJson.redirected;
    this.status = responseJson.status;
    this.statusText = responseJson.statusText;
    this.type = responseJson.type;
    this.url = responseJson.url;
    this.headers = new Headers(responseJson.headers);
  }

  json(): Promise<any> {
    return this.responseHandle.evaluate((response) => response.json());
  }

  text(): Promise<string> {
    return this.responseHandle.evaluate((response) => response.text());
  }

  get body(): NodeJS.ReadableStream | null {
    throw new Error("Cannot access body of FoodstuffsResponse directly");
  }
  get size(): number {
    throw new Error("Cannot access size of FoodstuffsResponse directly");
  }
  buffer(): Promise<Buffer> {
    throw new Error("Method not implemented.");
  }
  clone(): NodeResponse {
    throw new Error("Method not implemented.");
  }
  arrayBuffer(): Promise<ArrayBuffer> {
    throw new Error("Method not implemented.");
  }
  blob(): Promise<Blob> {
    throw new Error("Method not implemented.");
  }
  formData(): Promise<FormData> {
    throw new Error("Method not implemented.");
  }
}

const UNSERIALISABLE_RESPONSE_PROPS = [
  "arrayBuffer",
  "blob",
  "formData",
  "json",
  "text",
  "clone",
  "body",
] as const;

const SERIALISABLE_RESPONSE_PROPS = [
  "headers",
  "ok",
  "redirected",
  "status",
  "statusText",
  "type",
  "url",
] as const;

interface SerialisedResponse
  extends Omit<Response, typeof UNSERIALISABLE_RESPONSE_PROPS[number] | "headers"> {
  headers: Record<string, string>;
}
