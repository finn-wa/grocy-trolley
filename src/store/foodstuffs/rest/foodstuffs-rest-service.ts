import { RestService } from "@gt/utils/rest";
import { PAKNSAVE_URL } from "../models";
import { FoodstuffsAuthHeaderProvider } from "./foodstuffs-auth-header-provider";

export abstract class FoodstuffsRestService extends RestService {
  protected readonly baseUrl = this.validateBaseUrl(`${PAKNSAVE_URL}/CommonApi`);

  constructor(protected readonly authHeaderProvider: FoodstuffsAuthHeaderProvider) {
    super();
  }

  protected authHeaders() {
    return this.authHeaderProvider.authHeaders();
  }
}
