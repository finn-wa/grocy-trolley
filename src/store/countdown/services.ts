import { getEnvAs } from "@gt/utils/environment";
import { getBrowser } from "../shared/rest/browser";
import { CountdownOrderService } from "./orders/countdown-order-service";
import { CountdownUserAgent } from "./rest/countdown-user-agent";
import { CountdownTrolleyService } from "./trolley/countdown-trolley-service";

export async function countdownServices(): Promise<CountdownServices> {
  const loginDetails = getEnvAs({ COUNTDOWN_EMAIL: "email", COUNTDOWN_PASSWORD: "password" });
  const userAgent = new CountdownUserAgent(getBrowser, loginDetails);
  return {
    userAgent,
    trolleyService: new CountdownTrolleyService(userAgent),
    orderService: new CountdownOrderService(userAgent),
  };
}

export interface CountdownServices {
  userAgent: CountdownUserAgent;
  trolleyService: CountdownTrolleyService;
  orderService: CountdownOrderService;
}
