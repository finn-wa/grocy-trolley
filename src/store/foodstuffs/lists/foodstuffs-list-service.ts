import { AppTokens } from "@gt/app/di";
import { PromptProvider } from "@gt/prompts/prompt-provider";
import { uniqueByProperty } from "@gt/utils/arrays";
import { Logger } from "@gt/utils/logger";
import { setTimeout } from "timers/promises";
import { inject, Lifecycle, scoped } from "tsyringe";
import { FoodstuffsBaseProduct, PAKNSAVE_URL } from "../models";
import { FoodstuffsAuthHeaderProvider } from "../rest/foodstuffs-auth-header-provider";
import { FoodstuffsRestService } from "../rest/foodstuffs-rest-service";
import {
  formatListProductRef,
  List,
  ListProductRef,
  ListUpdate,
  toListProductRef,
} from "./foodstuffs-list.model";

export const TEMP_LIST_PREFIX = "[temporary]";

@scoped(Lifecycle.ContainerScoped)
export class FoodstuffsListService extends FoodstuffsRestService {
  protected readonly baseUrl = this.validateBaseUrl(`${PAKNSAVE_URL}/CommonApi`);
  protected readonly logger = new Logger(this.constructor.name);

  constructor(
    userAgent: FoodstuffsAuthHeaderProvider,
    @inject(AppTokens.promptProvider) private readonly prompt: PromptProvider
  ) {
    super(userAgent);
  }

  async createList(name: string): Promise<List> {
    const headersBuilder = await this.authHeaders();
    return this.putAndParse(this.buildUrl("ShoppingLists/CreateList", { name }), {
      headers: headersBuilder.acceptJson().build(),
    });
  }

  async createListWithNamePrompt(): Promise<List | null> {
    const name = await this.prompt.text("Enter a name for your new list");
    if (!name) return null;
    return this.createList(name);
  }

  async getLists(): Promise<List[]> {
    const headersBuilder = await this.authHeaders();
    return this.getAndParse<{ lists: List[] }>(this.buildUrl("ShoppingLists/GetLists"), {
      headers: headersBuilder.acceptJson().build(),
    }).then((res) => res.lists);
  }

  async getList(id: string): Promise<List> {
    const headersBuilder = await this.authHeaders();
    return this.getAndParse(this.buildUrl("ShoppingLists/GetList", { id }), {
      headers: headersBuilder.acceptJson().build(),
    });
  }

  /**
   * Prompts to select or create a new list
   * @returns the selected list's ID or null if the user cancelled the prompt
   */
  async promptSelectOrCreateList(): Promise<string | null> {
    const lists = await this.getLists();
    const listId = await this.prompt.select("Select list", [
      { title: "Create new list", value: "createNewList" as const },
      ...lists.map((list) => ({ title: list.name, value: list.listId })),
    ]);
    if (listId === "createNewList") {
      return this.createListWithNamePrompt().then((list) => list?.listId ?? null);
    }
    return listId;
  }

  /**
   * Performs a POST request to update the list with the specified ID. List
   * contents are replaced with the products in the listUpdate. To add products
   * to a list, use {@link addProductsToList}.
   * @param listUpdate Request body containing list ID and new products
   * @returns Updated list
   */
  async updateList(listUpdate: ListUpdate): Promise<Omit<List, "name">> {
    const headersBuilder = await this.authHeaders();
    return this.postAndParse(this.buildUrl("ShoppingLists/UpdateList"), {
      headers: headersBuilder.contentTypeJson().acceptJson().build(),
      body: JSON.stringify(listUpdate),
    });
  }

  async deleteList(id: string | number): Promise<{ lists: Omit<List, "products">[] }> {
    const headersBuilder = await this.authHeaders();
    return this.deleteAndParse(this.buildUrl(`ShoppingLists/DeleteList/${id}`), {
      headers: headersBuilder.acceptJson().build(),
    });
  }

  async createListWithProducts(
    name: string,
    products: ListProductRef[]
  ): Promise<Omit<List, "name">> {
    const createdList = await this.createList(name);
    return this.addProductsToList(createdList.listId, products);
  }

  async addProductsToList(listId: string, productsToAdd: ListProductRef[]): Promise<List> {
    const list = await this.getList(listId);
    const existingProducts = list.products;
    const products = uniqueByProperty(
      [...existingProducts.map(toListProductRef), ...productsToAdd.map(formatListProductRef)],
      "productId"
    );
    return this.updateList({ listId, products }) as Promise<List>;
  }

  async refreshProductPrices<T extends FoodstuffsBaseProduct>(products: T[]) {
    const list = await this.createListWithProducts(
      `${TEMP_LIST_PREFIX} price refresh - ${new Date().toISOString()}`,
      products.map((product) => toListProductRef(product))
    );
    const refreshedProducts = products.map((product) => {
      const listProduct = list.products.find((x) => x.productId === product.productId);
      if (!listProduct) {
        this.logger.warn(
          `Failed to find updated price for product: ${product.productId} / ${product.name}`
        );
        return { ...product };
      }
      return { ...product, price: listProduct?.price };
    });
    await this.deleteList(list.listId);
    return refreshedProducts;
  }

  /**
   * Deletes lists with names matching the supplied pattern.
   * @param namePattern Pattern to match list names
   * @param lists Optional array of lists to search for names in
   * @returns Array of responses to DELETE requests
   */
  async deleteLists(namePattern?: RegExp, lists?: List[]): Promise<Response[]> {
    if (!lists) {
      lists = await this.getLists();
    }
    const headersBuilder = await this.authHeaders();
    const request: RequestInit = { method: "DELETE", headers: headersBuilder.acceptJson().build() };
    const deletionUrls = lists
      .filter((list) => namePattern?.test(list.name) ?? true)
      .map(({ listId }) => this.buildUrl(`ShoppingLists/DeleteList/${listId}`));
    // Promise.all doesn't seem to work well
    const responses = [];
    for (const url of deletionUrls) {
      responses.push(await this.fetch(url, request));
      await setTimeout(250);
    }
    return responses;
  }
}
