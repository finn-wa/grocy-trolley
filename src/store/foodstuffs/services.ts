import { getEnvAs } from "@gt/utils/environment";
import { firefox } from "playwright";
import { LogLevel, playwrightLogger } from "@gt/utils/logger";
import { FoodstuffsCartService } from "./cart/foodstuffs-cart-service";
import { FoodstuffsListService } from "./lists/foodstuffs-list-service";
import { FoodstuffsOrderService } from "./orders/foodstuffs-order-service";
import { FoodstuffsSearchService } from "./search/foodstuffs-search-service";
import { FoodstuffsUserAgent } from "./rest/foodstuffs-user-agent";

export async function foodstuffsServices(): Promise<FoodstuffsServices> {
  const browser = await firefox.launch({
    headless: true,
    logger: playwrightLogger(LogLevel.WARN),
  });
  const userAgent = new FoodstuffsUserAgent(
    browser,
    getEnvAs({ PAKNSAVE_EMAIL: "email", PAKNSAVE_PASSWORD: "password" })
  );
  return {
    userAgent,
    cartService: new FoodstuffsCartService(userAgent),
    listService: new FoodstuffsListService(userAgent),
    orderService: new FoodstuffsOrderService(userAgent),
    searchService: new FoodstuffsSearchService(userAgent),
  };
}

export interface FoodstuffsServices {
  userAgent: FoodstuffsUserAgent;
  cartService: FoodstuffsCartService;
  listService: FoodstuffsListService;
  orderService: FoodstuffsOrderService;
  searchService: FoodstuffsSearchService;
}
