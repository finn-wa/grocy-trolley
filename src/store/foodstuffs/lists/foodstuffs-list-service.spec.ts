import { LoginDetails } from "@gt/store/shared/rest/login-details.model";
import { getEnvAs, initEnv } from "@gt/utils/environment";
import { FoodstuffsUserAgent } from "../rest/foodstuffs-user-agent";
import { getBrowser } from "../services";
import { FoodstuffsListService } from "./foodstuffs-list-service";
import { List, ListProductRef } from "./foodstuffs-list.model";

class ProductRefs {
  get milk(): ListProductRef {
    return { productId: "5201479-EA-000", quantity: 1, saleType: "UNITS" };
  }

  get carrots(): ListProductRef {
    return { productId: "5039965-KGM-000", quantity: 500, saleType: "WEIGHT" };
  }

  get bread(): ListProductRef {
    return { productId: "5011078-EA-000", quantity: 1, saleType: "UNITS" };
  }
  get manyItems(): ListProductRef[] {
    // unformatted: 5025751_EA_000PNS
    return [
      { productId: "5025751-EA-000", saleType: "UNITS", quantity: 1 },
      { productId: "5040543-EA-000", saleType: "UNITS", quantity: 1 },
      { productId: "5046503-EA-000", saleType: "UNITS", quantity: 1 },
      { productId: "5034241-EA-000", saleType: "UNITS", quantity: 1 },
      { productId: "5001241-EA-000", saleType: "UNITS", quantity: 1 },
      { productId: "5269328-EA-000", saleType: "UNITS", quantity: 1 },
      { productId: "5039965-KGM-000", saleType: "BOTH", quantity: 1 },
      { productId: "5039995-KGM-000", saleType: "BOTH", quantity: 1 },
      { productId: "5294812-EA-000", saleType: "UNITS", quantity: 1 },
      { productId: "5010008-EA-000", saleType: "UNITS", quantity: 1 },
      { productId: "5045921-EA-000", saleType: "UNITS", quantity: 1 },
      { productId: "5006357-EA-000", saleType: "UNITS", quantity: 1 },
      { productId: "5018301-EA-000", saleType: "UNITS", quantity: 1 },
      { productId: "5017803-EA-000", saleType: "UNITS", quantity: 1 },
    ];
  }
}

describe("FoodstuffsListService", () => {
  let listService: FoodstuffsListService;
  const refs = new ProductRefs();

  initEnv({
    envFilePath: ".test.env",
    envFilePathOptional: true,
    requiredVars: ["PAKNSAVE_EMAIL", "PAKNSAVE_PASSWORD"],
  });
  const loginDetails: LoginDetails = getEnvAs({
    PAKNSAVE_EMAIL: "email",
    PAKNSAVE_PASSWORD: "password",
  });
  const productIdsOf = (products: { productId: string }[]) =>
    products.map((p) => p.productId).sort();
  const expectArrayOfLength = (arr: ArrayLike<unknown>, length: number) =>
    expect(arr).toMatchObject<ArrayLike<unknown>>({ length });

  beforeEach(async () => {
    const userAgent = new FoodstuffsUserAgent(getBrowser, loginDetails);
    listService = new FoodstuffsListService(userAgent);
    await listService.deleteLists(/^(?!My Favourites)/);
  });

  describe("unit tests", () => {
    test("getLists", async () => {
      const lists = await listService.getLists();
      expectArrayOfLength(lists, 1);
      expect(lists[0]).toMatchObject({ name: "My Favourites", products: [] });
    });

    test("createList", async () => {
      const name = "strawberry";
      const list = await listService.createList(name);
      expect(list.name).toEqual(name);
      expect(list.products).toEqual([]);
      expect(list.listId).toBeTruthy();
    });

    test("createListWithProducts", async () => {
      const name = "raspberry";
      const products = refs.manyItems;
      const list = await listService.createListWithProducts(name, products);
      expect(list.listId).toBeTruthy();
      expect(productIdsOf(list.products)).toEqual(productIdsOf(products));
    });

    test("updateList", async () => {
      const name = "lemon";
      const { listId } = await listService.createList(name);
      const products = [refs.bread, refs.carrots];
      const list = await listService.updateList({ listId, products });
      expect(list.listId).toEqual(listId);
      expect(productIdsOf(list.products)).toEqual(productIdsOf(products));
    });

    test("deleteList", async () => {
      const name = "lime";
      const { listId } = await listService.createList(name);
      expect(await listService.getList(listId)).toBeTruthy();
      const response = await listService.deleteList(listId);
      expectArrayOfLength(response.lists, 1);
      expect(() => listService.getList(listId)).rejects.toThrowError(/404/);
    });

    test("deleteLists", async () => {
      for (const name of ["temporary list", "important items", "apr 1st (TEMP)"]) {
        await listService.createList(name);
      }
      const listsBefore = await listService.getLists();
      expectArrayOfLength(listsBefore, 4);
      const importantList = listsBefore.find((product) => product.name.includes("important"));
      expect(importantList).toBeTruthy();

      await listService.deleteLists(/temp/gi);
      const listsAfter = await listService.getLists();
      expectArrayOfLength(listsAfter, 2);
      expect(listsAfter).toContainEqual(importantList);
    });

    test("addProductsToList", async () => {
      const name = "persimmon";
      const { listId } = await listService.createList(name);
      expect(await listService.getList(listId)).toBeTruthy();
      await listService.updateList({ listId, products: [refs.milk] });

      const itemsToAdd = refs.manyItems;
      const updated = await listService.addProductsToList(listId, itemsToAdd);
      expectArrayOfLength(updated.products, itemsToAdd.length + 1);
    });
  });

  describe("snapshot tests", () => {
    function formatListForSnapshot(list: List): Partial<List> {
      return {
        ...list,
        listId: list.listId ? "uuid" : undefined,
        products: list.products?.map((product) => ({ ...product, price: 0 })),
      };
    }

    test("createList", async () => {
      const list = await listService.createList("orange");
      expect(formatListForSnapshot(list)).toMatchSnapshot();
    });

    test("getLists", async () => {
      const lists = await listService.getLists();
      expect(lists.map(formatListForSnapshot)).toMatchSnapshot();
    });

    test("getList", async () => {
      const { listId } = await listService.createListWithProducts("blueberry", [
        refs.bread,
        refs.carrots,
      ]);
      const list = await listService.getList(listId);
      expect(formatListForSnapshot(list)).toMatchSnapshot();
    });

    test("updateList", async () => {
      const { listId } = await listService.createListWithProducts("breadfruit", [refs.bread]);
      const list = await listService.updateList({ listId, products: [refs.carrots, refs.milk] });
      expect(formatListForSnapshot(list as List)).toMatchSnapshot();
    });

    test("deleteList", async () => {
      await listService.createListWithProducts("broccoli", [refs.milk]);
      const { listId } = await listService.createListWithProducts("tomato", [refs.bread]);
      const response = await listService.deleteList(listId);
      expect({
        ...response,
        lists: (response.lists as List[])?.map(formatListForSnapshot),
      }).toMatchSnapshot();
    });
  });
});
