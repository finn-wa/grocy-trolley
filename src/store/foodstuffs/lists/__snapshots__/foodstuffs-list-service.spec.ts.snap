// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoodstuffsListService snapshot tests createList 1`] = `
Object {
  "listId": "uuid",
  "name": "orange",
  "products": Array [],
}
`;

exports[`FoodstuffsListService snapshot tests deleteList 1`] = `
Object {
  "lists": Array [
    Object {
      "listId": "uuid",
      "name": "My Favourites",
      "products": undefined,
    },
    Object {
      "listId": "uuid",
      "name": "broccoli",
      "products": undefined,
    },
  ],
}
`;

exports[`FoodstuffsListService snapshot tests getList 1`] = `
Object {
  "listId": "uuid",
  "name": "blueberry",
  "originStatement": "Product origins may have changed since adding them to your list.",
  "products": Array [
    Object {
      "brand": "Ploughmans Bakery",
      "category": "Bakery",
      "imageUrl": "https://a.fsimg.co.nz/product/retail/fan/image/100x100/5011078.png",
      "liquor": false,
      "name": "Soy & Canterbury Linseed Bread",
      "price": 0,
      "productId": "5011078-EA-000",
      "promoBadgeImageTitle": "Extra Low",
      "promotionId": "124782404",
      "quantity": 1,
      "rangedInStore": true,
      "rangedOnline": true,
      "restricted": false,
      "saleTypes": Array [
        Object {
          "minUnit": 1,
          "stepSize": 1,
          "type": "UNITS",
          "unit": "ea",
        },
      ],
      "sale_type": "UNITS",
      "tobacco": false,
      "trolleyBadgeImageUrl": "PNS/6000-Extra_Low.svg",
      "weight": "",
      "weightUnitOfMeasure": "",
    },
    Object {
      "category": "Fruit & Vegetables",
      "imageUrl": "https://a.fsimg.co.nz/product/retail/fan/image/100x100/5039965.png",
      "liquor": false,
      "name": "Carrots",
      "price": 0,
      "productId": "5039965-KGM-000",
      "promoBadgeImageTitle": "Everyday Low",
      "quantity": 500,
      "rangedInStore": true,
      "rangedOnline": true,
      "restricted": false,
      "saleTypes": Array [
        Object {
          "minUnit": 200,
          "stepSize": 100,
          "type": "WEIGHT",
          "unit": "g",
        },
      ],
      "sale_type": "WEIGHT",
      "tobacco": false,
      "trolleyBadgeImageUrl": "PNS/Everyday_Low.svg",
      "weight": "150",
      "weightUnitOfMeasure": "g",
    },
  ],
}
`;

exports[`FoodstuffsListService snapshot tests getLists 1`] = `
Array [
  Object {
    "listId": "uuid",
    "name": "My Favourites",
    "products": Array [],
  },
]
`;

exports[`FoodstuffsListService snapshot tests updateList 1`] = `
Object {
  "listId": "uuid",
  "products": Array [
    Object {
      "category": "Bakery",
      "imageUrl": "https://a.fsimg.co.nz/product/retail/fan/image/100x100/5011078.png",
      "liquor": false,
      "name": "Soy & Canterbury Linseed Bread",
      "price": 0,
      "productId": "5011078-EA-000",
      "quantity": 1,
      "rangedInStore": true,
      "rangedOnline": true,
      "restricted": false,
      "saleTypes": Array [
        Object {
          "minUnit": 1,
          "stepSize": 1,
          "type": "UNITS",
          "unit": "ea",
        },
      ],
      "sale_type": "UNITS",
      "tobacco": false,
      "weight": "",
      "weightUnitOfMeasure": "",
    },
    Object {
      "category": "Fruit & Vegetables",
      "imageUrl": "https://a.fsimg.co.nz/product/retail/fan/image/100x100/5039965.png",
      "liquor": false,
      "name": "Carrots",
      "price": 0,
      "productId": "5039965-KGM-000",
      "quantity": 500,
      "rangedInStore": true,
      "rangedOnline": true,
      "restricted": false,
      "saleTypes": Array [
        Object {
          "minUnit": 200,
          "stepSize": 100,
          "type": "WEIGHT",
          "unit": "g",
        },
      ],
      "sale_type": "WEIGHT",
      "tobacco": false,
      "weight": "150",
      "weightUnitOfMeasure": "g",
    },
    Object {
      "category": "Dairy & Eggs",
      "imageUrl": "https://a.fsimg.co.nz/product/retail/fan/image/100x100/5201479.png",
      "liquor": false,
      "name": "Standard Milk",
      "price": 0,
      "productId": "5201479-EA-000",
      "quantity": 1,
      "rangedInStore": true,
      "rangedOnline": true,
      "restricted": false,
      "saleTypes": Array [
        Object {
          "minUnit": 1,
          "stepSize": 1,
          "type": "UNITS",
          "unit": "ea",
        },
      ],
      "sale_type": "UNITS",
      "tobacco": false,
      "weight": "",
      "weightUnitOfMeasure": "",
    },
  ],
}
`;

exports[`FoodstuffsListService unit tests createList 1`] = `
Object {
  "listId": "uuid",
  "name": "orange",
  "products": Array [],
}
`;

exports[`FoodstuffsListService unit tests deleteList 1`] = `
Object {
  "lists": Array [
    Object {
      "listId": "uuid",
      "name": "My Favourites",
      "products": undefined,
    },
    Object {
      "listId": "uuid",
      "name": "broccoli",
      "products": undefined,
    },
  ],
}
`;

exports[`FoodstuffsListService unit tests getList 1`] = `
Object {
  "listId": "uuid",
  "name": "blueberry",
  "originStatement": "Product origins may have changed since adding them to your list.",
  "products": Array [
    Object {
      "brand": "Ploughmans Bakery",
      "category": "Bakery",
      "imageUrl": "https://a.fsimg.co.nz/product/retail/fan/image/100x100/5011078.png",
      "liquor": false,
      "name": "Soy & Canterbury Linseed Bread",
      "price": 0,
      "productId": "5011078-EA-000",
      "promoBadgeImageTitle": "Extra Low",
      "promotionId": "124782404",
      "quantity": 1,
      "rangedInStore": true,
      "rangedOnline": true,
      "restricted": false,
      "saleTypes": Array [
        Object {
          "minUnit": 1,
          "stepSize": 1,
          "type": "UNITS",
          "unit": "ea",
        },
      ],
      "sale_type": "UNITS",
      "tobacco": false,
      "trolleyBadgeImageUrl": "PNS/6000-Extra_Low.svg",
      "weight": "",
      "weightUnitOfMeasure": "",
    },
    Object {
      "category": "Fruit & Vegetables",
      "imageUrl": "https://a.fsimg.co.nz/product/retail/fan/image/100x100/5039965.png",
      "liquor": false,
      "name": "Carrots",
      "price": 0,
      "productId": "5039965-KGM-000",
      "promoBadgeImageTitle": "Everyday Low",
      "quantity": 500,
      "rangedInStore": true,
      "rangedOnline": true,
      "restricted": false,
      "saleTypes": Array [
        Object {
          "minUnit": 200,
          "stepSize": 100,
          "type": "WEIGHT",
          "unit": "g",
        },
      ],
      "sale_type": "WEIGHT",
      "tobacco": false,
      "trolleyBadgeImageUrl": "PNS/Everyday_Low.svg",
      "weight": "150",
      "weightUnitOfMeasure": "g",
    },
  ],
}
`;

exports[`FoodstuffsListService unit tests getLists 1`] = `
Array [
  Object {
    "listId": "uuid",
    "name": "My Favourites",
    "products": Array [],
  },
]
`;

exports[`FoodstuffsListService unit tests updateList 1`] = `
Object {
  "listId": "uuid",
  "products": Array [
    Object {
      "category": "Bakery",
      "imageUrl": "https://a.fsimg.co.nz/product/retail/fan/image/100x100/5011078.png",
      "liquor": false,
      "name": "Soy & Canterbury Linseed Bread",
      "price": 0,
      "productId": "5011078-EA-000",
      "quantity": 1,
      "rangedInStore": true,
      "rangedOnline": true,
      "restricted": false,
      "saleTypes": Array [
        Object {
          "minUnit": 1,
          "stepSize": 1,
          "type": "UNITS",
          "unit": "ea",
        },
      ],
      "sale_type": "UNITS",
      "tobacco": false,
      "weight": "",
      "weightUnitOfMeasure": "",
    },
    Object {
      "category": "Fruit & Vegetables",
      "imageUrl": "https://a.fsimg.co.nz/product/retail/fan/image/100x100/5039965.png",
      "liquor": false,
      "name": "Carrots",
      "price": 0,
      "productId": "5039965-KGM-000",
      "quantity": 500,
      "rangedInStore": true,
      "rangedOnline": true,
      "restricted": false,
      "saleTypes": Array [
        Object {
          "minUnit": 200,
          "stepSize": 100,
          "type": "WEIGHT",
          "unit": "g",
        },
      ],
      "sale_type": "WEIGHT",
      "tobacco": false,
      "weight": "150",
      "weightUnitOfMeasure": "g",
    },
    Object {
      "category": "Dairy & Eggs",
      "imageUrl": "https://a.fsimg.co.nz/product/retail/fan/image/100x100/5201479.png",
      "liquor": false,
      "name": "Standard Milk",
      "price": 0,
      "productId": "5201479-EA-000",
      "quantity": 1,
      "rangedInStore": true,
      "rangedOnline": true,
      "restricted": false,
      "saleTypes": Array [
        Object {
          "minUnit": 1,
          "stepSize": 1,
          "type": "UNITS",
          "unit": "ea",
        },
      ],
      "sale_type": "UNITS",
      "tobacco": false,
      "weight": "",
      "weightUnitOfMeasure": "",
    },
  ],
}
`;
